### PR TITLE
Rewrite `ConditionalBindingCascadeRule`

### LINKED ISSUE #617: Conditional binding cascade rule does not work
Dear team,

Below is my sample code

// MARK: Condition binding cascade rule
    func conditionBindingCascade() -> Void {
        let a: String? = "a"
        let b: String? = "b"
        
        // Preffered
        if let a = a, b = b {
            debugPrint("Preffered \(a),\(b)")
        }
        
        // Not preffered
        if let c = a, let d = b {
            debugPrint("Not preffered \(c),\(d)")
        }
    }

And it is not warned. This rule has enabled, I'd checked in command line with "swiftlint rules" command.

### PR DESCRIPTION
**Pull request by norio-nomura**
- Fix #617
- Add `ASTBaseRule`  
`ASTBaseRule` takes kind as `String` for supporting other than `SwiftDeclarationKind`.
`ASTRule` inherits `ASTBaseRule`.

### CODE REVIEW

**Comment by norio-nomura** [comment]
Updated.
- Add false negative example to ConditionalBindingCascadeRule from #617
- Add changelog entry

**Comment by norio-nomura** [comment]
`make sum_test` fails randomly on both Travis-CI and local mac.
We need to update SourceKitten in `Package.swift`, but I won't update it in this PR.

File: CHANGELOG.md [hunk 1]
```diff
@@ -15,6 +15,10 @@
 * SwiftLint no longer crashes when SourceKitService crashes.  
   [Norio Nomura](https://github.com/norio-nomura)
 
+* Rewrite `conditional_binding_cascade` rule.  
+  [Norio Nomura](https://github.com/norio-nomura)
+  [#617](https://github.com/jpsim/SwiftLint/issues/617)
```
Comment (jpsim) [inline CHANGELOG.md]: This is the wrong url

File: Tests/SwiftLintFramework/IntegrationTests.swift [hunk 2]
```diff
@@ -51,6 +51,7 @@ extension String {
             let rawPointer = $0._rawValue
             let byteSize = lengthOfBytesUsingEncoding(NSUTF8StringEncoding)._builtinWordValue
             let isASCII = true._getBuiltinLogicValue()
+            // swiftlint:disable:next variable_name
```
Comment (jpsim) [inline Tests/SwiftLintFramework/IntegrationTests.swift]: Why the disable here?

**Comment by jpsim** [comment]
:+1: other than where noted

File: Tests/SwiftLintFramework/IntegrationTests.swift (hunk 2 above)
Comment (norio-nomura) [inline Tests/SwiftLintFramework/IntegrationTests.swift]: I got `variable_name` violation several time about `_builtinStringLiteral` on `Run SwiftLint` build phase and  `IntegrationTests.testSwiftLintLints`.
But, now I can't reproduce that after doing `Clean` on Xcode. 🤔
I'll remove this.

Comment (norio-nomura) [inline Tests/SwiftLintFramework/IntegrationTests.swift]: I want to keep this comments.
I reverted instead of removed.

Comment (norio-nomura) [inline Tests/SwiftLintFramework/IntegrationTests.swift]: Test failed on Travis-CI https://travis-ci.org/realm/SwiftLint/builds/123487152 both of Xcode and SPM.
> SwiftLintFrameworkTests.IntegrationTests
  testSwiftLintLints, failed - Variable name should only contain alphanumeric characters: '_builtinStringLiteral'
  /Users/travis/build/realm/SwiftLint/Tests/SwiftLintFramework/IntegrationTests.swift:54
```
            let isASCII = true._getBuiltinLogicValue()
            let staticString = StaticString(_builtinStringLiteral: rawPointer, byteSize: byteSize,
                isASCII: isASCII)
```

Reverted again.

**Comment by norio-nomura** [comment]
Performance note:
The duration of `ConditionalBindingCascadeRule` on linting Carthage-16.2 increased from 64ms (w/ 62c7d4a99) to 283ms (w/ d3f698f).

**Comment by jpsim** [comment]
Looks great to me, thanks!

**Comment by norio-nomura** [comment]
Thanks for reviews and merging! 🙏

**Pull request merged**